25.Blur/main.cpp

```diff
@@ -4,182 +4,500 @@
+
+        driver->bindDescriptorSets(video::EPBP_COMPUTE, pipeline_horizontal->getLayout(), 0u, 1u, &ds_horizontal.get(), nullptr);
+        driver->bindComputePipeline(pipeline_horizontal.get());
+        BlurClass::dispatchHelper(driver, pipeline_horizontal->getLayout(), pushConstants[0], dispatchInfo[0]);
+
+        driver->bindDescriptorSets(video::EPBP_COMPUTE, pipeline_vertical->getLayout(), 0u, 1u, &ds_vertical.get(), nullptr);
+        driver->bindComputePipeline(pipeline_vertical.get());
+        BlurClass::dispatchHelper(driver, pipeline_vertical->getLayout(), pushConstants[1], dispatchInfo[1], false);
+
+        video::COpenGLExtensionHandler::extGlMemoryBarrier(GL_FRAMEBUFFER_BARRIER_BIT);
+
+        driver->blitRenderTargets(blit_fbo, nullptr, false, false);
+
+        driver->endScene();
+    }
 
-    device->drop();
+    if (!ext::ScreenShot::createScreenShot(driver, am, out_image_view.get(), "../screenshot.png", asset::EF_R8G8B8_SRGB))
+        std::cout << "Unable to create screenshot" << std::endl;
 
     return 0;
 }
+#endif
+
+class BlurTestApp : public ApplicationBase
+{
+    constexpr static inline uint32_t FRAMES_IN_FLIGHT = 5u;
+	constexpr static inline uint32_t SC_IMG_COUNT = 3u;
+	constexpr static inline uint64_t MAX_TIMEOUT = 99999999999999ull;
+
+public:
+	void onAppInitialized_impl() override
+	{
+		CommonAPI::InitOutput initOutput;
+        CommonAPI::InitWithDefaultExt(initOutput, video::EAT_VULKAN, "Blur", FRAMES_IN_FLIGHT, 1024, 1024, SC_IMG_COUNT, asset::IImage::EUF_COLOR_ATTACHMENT_BIT);
+
+		system = std::move(initOutput.system);
+		window = std::move(initOutput.window);
+		windowCb = std::move(initOutput.windowCb);
+		apiConnection = std::move(initOutput.apiConnection);
+		surface = std::move(initOutput.surface);
+		physicalDevice = std::move(initOutput.physicalDevice);
+		logicalDevice = std::move(initOutput.logicalDevice);
+		utilities = std::move(initOutput.utilities);
+		queues = std::move(initOutput.queues);
+		swapchain = std::move(initOutput.swapchain);
+		commandPools = std::move(initOutput.commandPools);
+		assetManager = std::move(initOutput.assetManager);
+		cpu2gpuParams = std::move(initOutput.cpu2gpuParams);
+		logger = std::move(initOutput.logger);
+		inputSystem = std::move(initOutput.inputSystem);
+
+	}
+
+	void onAppTerminated_impl() override
+	{
+		logicalDevice->waitIdle();
+	}
+
+	void workLoopBody() override
+	{
+        inputSystem->getDefaultKeyboard(&keyboard);
+        keyboard.consumeEvents(
+            [this](const ui::IKeyboardEventChannel::range_t& events)
+            {
+                for (auto eventIt = events.begin(); eventIt != events.end(); ++eventIt)
+                {
+                    const auto& ev = *eventIt;
+                    if ((ev.keyCode == ui::EKC_Q) && (ev.action == ui::SKeyboardEvent::ECA_RELEASED))
+                    {
+                        m_appRunning = false;
+                    }
+                }
+            },
+            logger.get());
```

Comment:
give up on Q-to-quit, just handle the window close event properly

Reply:
just use `keepRunning`

Reply:
Removed Q-to-quit, [but I was using `keepRunning`, wasn't I?](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/blob/a7cced000ac01905d158206a6dad6d1ec834a9dc/25.Blur/main.cpp#L524)

Reply:
yes, good.